pkg/querier/merge_queryable.go

```diff
@@ -0,0 +1,280 @@
+func (m *addLabelsSeriesSet) Err() error {
+	return m.upstream.Err()
+}
+
+// A collection of warnings for the whole set.
+// Warnings could be return even iteration has not failed with error.
+func (m *addLabelsSeriesSet) Warnings() storage.Warnings {
+	return m.upstream.Warnings()
+}
+
+type addLabelsSeries struct {
+	upstream storage.Series
+	labels   labels.Labels
+}
+
+// Labels returns the complete set of labels. For series it means all labels identifying the series.
+func (a *addLabelsSeries) Labels() labels.Labels {
+	return setLabelsRetainExisting(a.upstream.Labels(), a.labels...)
+}
+
+// Iterator returns a new, independent iterator of the data of the series.
+func (a *addLabelsSeries) Iterator() chunkenc.Iterator {
+	return a.upstream.Iterator()
+}
+
+// Querier returns a new mergeQuerier aggregating all tenants into the result
+func (m *mergeQueryable) Querier(ctx context.Context, mint int64, maxt int64) (storage.Querier, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(tenantIDs) <= 1 {
+		return m.upstream.Querier(ctx, mint, maxt)
+	}
+
+	var queriers = make([]storage.Querier, len(tenantIDs))
+	for pos, tenantID := range tenantIDs {
+		q, err := m.upstream.Querier(
+			user.InjectOrgID(ctx, tenantID),
+			mint,
+			maxt,
+		)
+		if err != nil {
+			return nil, err
+		}
+		queriers[pos] = q
+	}
+
+	return &mergeQuerier{
+		queriers:        queriers,
+		tenantIDs:       tenantIDs,
+		tenantLabelName: defaultTenantLabel,
+	}, nil
+}
+
+const retainExistingPrefix = "original_"
+
+// this sets a label and preserves an existing value a new label prefixed with
+// original_. It doesn't do this recursively.
+func setLabelRetainExisting(ls labels.Labels, lb *labels.Builder, additional labels.Label) {
+	if oldValue := ls.Get(additional.Name); oldValue != "" {
+		lb.Set(
+			retainExistingPrefix+additional.Name,
+			oldValue,
+		)
+	}
+	lb.Set(additional.Name, additional.Value)
+}
+
+func setLabelsRetainExisting(src labels.Labels, additionalLabels ...labels.Label) labels.Labels {
+	lb := labels.NewBuilder(src)
+
+	for _, additionalL := range additionalLabels {
+		setLabelRetainExisting(src, lb, additionalL)
+	}
+
+	return lb.Labels()
+}
+
+type stringSliceFunc func() ([]string, storage.Warnings, error)
+
+func mergeDistinctStringSlice(funcs ...stringSliceFunc) ([]string, storage.Warnings, error) {
+	if len(funcs) == 1 {
+		return funcs[0]()
+	}
+
+	var warnings storage.Warnings
+	resultMap := make(map[string]struct{})
+	for _, f := range funcs {
+		result, w, err := f()
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, e := range result {
+			resultMap[e] = struct{}{}
+		}
+		warnings = append(warnings, w...)
+	}
+
+	var result []string
+	for e := range resultMap {
+		result = append(result, e)
+	}
+	sort.Strings(result)
+	return result, warnings, nil
+}
+
+func filterValuesByMatchers(labelName string, labelValues []string, matchers ...*labels.Matcher) ([]int, []*labels.Matcher) {
```

Comment:
It took me a while to figure out what this function does and why it is needed. Could you add comment explaining it a bit more. It also may be a bit easier if instead of having a separate function you just put this logic into the`mergeQuerier.Select` function and comment it there. I don't think it has much potential for reuse and having it inline would make things a bit easier to follow.

Reply:
I move the method close to Select and mentioned it in the comment specifically. If you think I should move it in line I can still do that...